src/util/pelz_io.c

```diff
@@ -0,0 +1,411 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <openssl/bio.h>
+#include <openssl/evp.h>
+#include <openssl/buffer.h>
+
+#include "CharBuf.h"
+#include "pelz_log.h"
+#include "pelz_io.h"
+#include "key_table.h"
+#include "pelz_request_handler.h"
+#include "util.h"
+
+int get_file_ext(CharBuf buf, int *ext)
+{
+  int period_index = 0;
+  char *ext_value;
+  int ext_len = 0;
+  int ext_type_size = 3;
+  char *ext_type[3] = { ".txt", ".pem", ".key" };
+
+  period_index = getIndexForChar(buf, '.', (buf.len - 1), 1);
+  ext_len = (buf.len - period_index);
+  ext_value = calloc(ext_len, sizeof(char));
```

Comment:
I don't think this can currently cause an off-by-1 error because of commit 4a9bfd9

Currently, pelz only passes in null-terminated file paths, but this function does not indicate it is necessary to do so. This could become an issue if ext_value is expected to null-terminate. We could add comments to the header/here, but I think we should explicitly check to make sure ext_value is null-terminated.

Reply:
There is potential for a bug here, but it's slightly different. If `buf->chars` is `null` terminated and `buf->len` includes space for that `null` character then the `ext_len == strlen(ext_type[i])` check will always fail, since `ext_len` will be 1 too long.

The bug fixed in 4a9bfd9 happened because when the file path got packed into a `CharBuf` the allocated space (and computed length) didn't include the `null` terminator.

I'm inclined to leave this alone for now, since I like the current behavior (`CharBuf`'s don't [and shouldn't] carry a `null` terminator, they carry lengths) and the parsing here is based entirely on the length provided by `buf`.